File: mycollab-scheduler-ondemand/src/main/scala/com/mycollab/ondemand/schedule/jobs/DeleteObsoleteAccountJob.scala
package com.mycollab.ondemand.schedule.jobs

import com.google.common.eventbus.AsyncEventBus
import com.mycollab.db.arguments.{BasicSearchRequest, DateSearchField, SetSearchField}
import com.mycollab.module.billing.AccountStatusConstants
import com.mycollab.ondemand.module.billing.domain.criteria.BillingAccountSearchCriteria
import com.mycollab.ondemand.module.billing.esb.DeleteAccountEvent
import com.mycollab.ondemand.module.billing.service.BillingService
import com.mycollab.schedule.jobs.GenericQuartzJobBean
import org.joda.time.LocalDate
import org.quartz.{JobExecutionContext, JobExecutionException}
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.config.BeanDefinition
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Component

/**
  * @author MyCollab Ltd
  * @since 5.2.6
  */
@Component
@Scope(BeanDefinition.SCOPE_PROTOTYPE)
class DeleteObsoleteAccountJob extends GenericQuartzJobBean {
  @Autowired private val billingService: BillingService = null
  @Autowired private val asyncEventBus: AsyncEventBus = null

  @throws(classOf[JobExecutionException])
  def executeJob(context: JobExecutionContext): Unit = {
    val searchCriteria = new BillingAccountSearchCriteria
    searchCriteria.setStatuses(new SetSearchField[String](AccountStatusConstants.TRIAL, AccountStatusConstants.INVALID))
    searchCriteria.setLastAccessTime(new DateSearchField(new LocalDate().minusDays(30).toDate))
    import collection.JavaConverters._
    val obsoleteAccounts = billingService.findPageableListByCriteria(new
        BasicSearchRequest[BillingAccountSearchCriteria](searchCriteria, 0, Integer.MAX_VALUE)).asScala.toList
    for (obsoleteAccount <- obsoleteAccounts) {
      val deleteAccountEvent = new DeleteAccountEvent(obsoleteAccount.getId, null)
      asyncEventBus.post(deleteAccountEvent)
    }
  }
}

File: mycollab-scheduler-ondemand/src/main/java/com/mycollab/ondemand/schedule/jobs/DeleteObsoleteAccountJob.kt
package com.mycollab.ondemand.schedule.jobs

import com.mycollab.db.arguments.BasicSearchRequest
import com.mycollab.db.arguments.DateSearchField
import com.mycollab.db.arguments.SetSearchField
import com.mycollab.module.billing.AccountStatusConstants
import com.mycollab.ondemand.module.billing.domain.criteria.BillingAccountSearchCriteria
import com.mycollab.ondemand.module.billing.service.BillingService
import com.mycollab.schedule.jobs.GenericQuartzJobBean
import org.joda.time.LocalDate
import org.quartz.JobExecutionContext
import org.quartz.JobExecutionException
import org.springframework.beans.factory.config.BeanDefinition
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Component

/**
 * @author MyCollab Ltd
 * @since 6.0.0
 */
@Component
@Scope(BeanDefinition.SCOPE_PROTOTYPE)
class DeleteObsoleteAccountJob(private val billingService: BillingService) : GenericQuartzJobBean() {

    @Throws(JobExecutionException::class)
    override fun executeJob(context: JobExecutionContext) {
        val searchCriteria = BillingAccountSearchCriteria()
        searchCriteria.statuses = SetSearchField(AccountStatusConstants.TRIAL)
        searchCriteria.lastAccessTime = DateSearchField(LocalDate().minusDays(30).toDate())

//        val obsoleteAccounts = billingService.findPageableListByCriteria(BasicSearchRequest(searchCriteria))
//        obsoleteAccounts.forEach { billingService.cancelAccount(it.id, null) }
    }
}